src/hotspot/cpu/x86/c2_stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,1850 @@
+    //  r15 = unused
+    //  XMM_BYTE_0 - first element of needle, broadcast
+    //  XMM_BYTE_K - last element of needle, broadcast
+    //
+    //  The haystack is <= 32 bytes
+    //
+    // If a match is not found, branch to L_returnRBP (which will always
+    // return -1).
+    //
+    // If a match is found, jump to L_checkRangeAndReturn, which ensures the
+    // matched needle is not past the end of the haystack.
+
+    const Register haystack = rbx;
+    const Register needle = r14;
+    const Register needle_val = r8;
+    const Register set_bit = r11;
+    const Register eq_mask = rsi;
+    const Register rTmp = rax;
+
+    for (int i = 6; i < NUMBER_OF_CASES; i++) {
+      small_hs_jmp_table[i] = __ pc();
+      if (isU && ((i + 1) & 1)) {
+        continue;
+      } else {
+        broadcast_additional_needles(true, i + 1, needle, noreg, rTmp, ae, _masm);
+
+        compare_haystack_to_needle(true, i + 1, L_error, haystack, eq_mask, noreg, rTmp, XMM_TMP1,
+                                   XMM_TMP2, ae, _masm);
+
+        byte_compare_helper(i + 1, L_error, L_checkRange, needle, needle_val, haystack, eq_mask,
+                            set_bit, rTmp, ae, _masm);
+      }
+    }
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////////////
+  //
+  // Large haystack (> 32 bytes) switch
+
+  {
+
+    ////////////////////////////////////////////////
+    //  On entry to each case of large_hs, the register state is:
+    //
+    //  rax = unused
+    //  rbx = &haystack
+    //  rcx = haystack length
+    //  rdx = &needle
+    //  rsi = haystack length
+    //  rdi = &haystack
+    //  rbp = -1
+    //  r8  = unused
+    //  r9  = unused
+    //  r10 = hs_len - needle len
+    //  r11 = unused
+    //  r12 = needle length
+    //  r13 = (needle length - 1)
+    //  r14 = &needle
+    //  r15 = unused
+    //  XMM_BYTE_0 - first element of needle, broadcast
+    //  XMM_BYTE_K - last element of needle, broadcast
+    //
+    //  The haystack is > 32 bytes
+
+    const Register haystack = rbx;
+    const Register needle = r14;
+    const Register needle_len = r12;
+    const Register needle_val = r15;
+    const Register set_bit = r8;
+    const Register eq_mask = r9;
+    const Register hs_ptr = rcx;
+    const Register hsLength = rsi;
+    const Register rTmp1 = rdi;
+    const Register rTmp2 = r15;
+    const Register rTmp3 = rdx;
+    const Register rTmp4 = r13;
+
+    for (int i = 0; i < NUMBER_OF_CASES; i++) {
+      big_hs_jmp_table[i] = __ pc();
+      if (isU && ((i + 1) & 1)) {
+        continue;
+      } else {
+        Label L_loopTop;
+
+        big_case_loop_helper(true, i + 1, L_checkRange, L_loopTop, eq_mask, hs_ptr, needle_len,
+                             needle, haystack, hsLength, rTmp1, rTmp2, rTmp3, rTmp4, ae, _masm);
+        byte_compare_helper(i + 1, L_loopTop, L_fixup, needle, needle_val, hs_ptr, eq_mask, set_bit,
+                            rTmp4, ae, _masm);
+      }
```

Comment:
L_checkRange on NoMatch could be L_error instead.

Reply:
Fixed